### PR TITLE
Implement graph, nodes and edge creation

### PR DESCRIPTION
**Pull request by codejaeger**
* GraphAnimation

## PR Checklist

If you are contributing to `Javis.jl`, please make sure you are able to check off each item on this list:

- [ ] Did I update `CHANGELOG.md` with whatever changes/features I added with this PR?
- [ ] Did I make sure to only change the part of the file where I introduced a new change/feature?
- [ ] Did I cover all corner cases to be close to 100% test coverage (if applicable)?
- [ ] Did I properly add Javis dependencies to the `Project.toml` + set an upper bound of the dependency (if applicable)?
- [ ] Did I properly add test dependencies to the `test` directory (if applicable)?
- [ ] Did I check relevant tutorials that may be affected by changes in this PR?
- [ ] Did I clearly articulate why this PR was made the way it was and how it was made?

**Link to relevant issue(s)**
Closes #

**How did you address these issues with this PR? What methods did you use?**

### CODE REVIEW

File: src/structs/Graphs.jl [hunk 1]
```diff
@@ -14,21 +41,50 @@ struct GraphAnimation
...
+    node_property_limits::Dict{Symbol,Tuple{Real,Real}}
+end
+
+"""
+    GraphAnimation(width::Int, height::Int, directed::Bool, [start_pos])
+
+Create a graph object with no nodes or edges yet.
+"""
+GraphAnimation(width::Int, height::Int, directed::Bool, start_pos::Union{Point,Object}=O) = directed ?
+    GraphAnimation(LightGraphs.SimpleDiGraph(), directed, width::Int, height::Int, start_pos) :
```
Comment (Wikunia) [inline src/structs/Graphs.jl:54]: no need to write `width::Int, height::Int` in the call

File: src/structs/Graphs.jl [hunk 2]
```diff
@@ -14,21 +41,50 @@ struct GraphAnimation
...
+    background("white")
+    sethue("black")
 end
 
-GraphAnimation(reference_graph, directed::bool) =
-    GraphAnimation(reference_graph, directed, 300, 300, O)
-GraphAnimation(width::Int, height::Int, start_pos::Union{Point,Object}) =
-    GraphAnimation(LightGraphs.SimpleGraph(), false, width::Int, height::Int, start_pos)
+video = Video(300, 300)
+g = GraphAnimation(100, 100, false, O)
```
Comment (Wikunia) [inline src/structs/Graphs.jl:71]: does not fit the syntax of the function as it specifies `graph`.

File: src/structs/Graphs.jl [hunk 3]
```diff
@@ -14,21 +41,50 @@ struct GraphAnimation
...
+    GraphAnimation(width::Int, height::Int, directed::Bool, [start_pos])
+
+Create a graph object with no nodes or edges yet.
+"""
+GraphAnimation(width::Int, height::Int, directed::Bool, start_pos::Union{Point,Object}=O) = directed ?
+    GraphAnimation(LightGraphs.SimpleDiGraph(), directed, width::Int, height::Int, start_pos) :
+    GraphAnimation(LightGraphs.SimpleGraph(), directed, width::Int, height::Int, start_pos)
+
+"""
+    GraphAnimation(graph, directed::Bool, width::Int, height::Int, start_pos::Union{Point,Object}; <keyword arguments>)
```
Comment (Wikunia) [inline src/structs/Graphs.jl:58]: what are those keyword arguments?

**Review by Wikunia — COMMENTED** [review]
Just some comments ;)

File: src/structs/JGraph.jl [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+                    )
+                end
+            end
+        end
+    end
+end
+
+function _global_layout(video, object, frames; kwargs...)
+    g = object.meta
+    layout_x = layout_y = []
```
Comment (Wikunia) [inline src/structs/JGraph.jl:170]: This doesn't work as you probably expect.
i.e: 
```julia
a = b = []
push!(a, 1)
```
will change `b`

File: src/structs/JGraph.jl [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+        end
+        # Normalize coordinates between -0.5 and 0.5
+        coords = map((p) -> Point(p), collect(zip(layout_x, layout_y))) .- [(0.5, 0.5)]
+        # Scale to graph dimensions
+        coords = coords .* [(g.width, g.height)]
+        object.opts[:layout] = coords
+    end
+    # Now assign positions back to all nodes
+    for (idx, p) in enumerate(node_props(g.adjacency_list))
+        g.ordering[p].meta.opts[:position] = object.opts[:layout][idx]
```
Comment (Wikunia) [inline src/structs/JGraph.jl:199]: is this used to give the position to each node later in rendering?

**Review by Wikunia — COMMENTED** [review]
Looks good. Just some questions before I go deeper :wink:
Maybe you can add some flow diagram/description of how everything is connected and give some details/docstrings to some more functions. Mostly describing what they are used for/doing.

File: src/Javis.jl [hunk 6]
```diff
@@ -102,6 +109,10 @@ include("javis_viewer.jl")
 include("latex.jl")
 include("object_values.jl")
 
+include("graph_animations.jl")
+include("node_shapes.jl")
+include("edge_shapes.jl")
```
Comment (Wikunia) [inline src/Javis.jl:114]: doesn't exist

File: examples/graph_animations/demo.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    if i%2 == 0
+        @Graph g i*10:150 GraphVertex(i, [node_shape(:circle, true, radius=12), node_fill(:image, "football.png"), node_text("$(i)", :top)]) O
+    else
+        @Graph g i*10:150 GraphVertex(i, [node_shape(:rectangle, true, width=20, height=20), node_fill(:color, "yellow"), node_text(L"""%$i""", :inside), node_border("green", 2)]) O
+    end
+end
+count = 0
+for i in 1:length(adjacency_list)
+    for j in adjacency_list[i]
+        @Graph g 15+count*10:150 GraphEdge(i, j, (args...; kwargs...)->edge(; kwargs...)) O
```
Comment (Wikunia) [inline examples/graph_animations/demo.md:39]: doesn't work with `count` as it's not known when you write this in global scope

File: src/node_shapes.jl [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+- `radius`: Specify if shape is `:circle`.
+- `width`: Specify if shape is `:rectangle`.
+- `height`: Specify if shape is `:rectangle`.
+"""
+function node_shape(shape::Symbol, clip::Bool=false; kwargs...)
+    kw = Dict(collect(kwargs)...)
+    opts = Dict{Symbol, Any}()
+    opts[:shape] = shape
+    opts[:clip] = clip
+    if shape == :rectangle
```
Comment (Wikunia) [inline src/node_shapes.jl:22]: currently there is no error message when one writes:
```julia
node_shape(:rectangle, true, radius=12)
```

Maybe have a look at: https://docs.julialang.org/en/v1/manual/types/#%22Value-types%22 to dispatch on the `Symbol`

File: src/node_shapes.jl [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+        shift = (O, O)
+        if text_align == :top
+            shift = ((0, -text_dims[2]-offset), (0, -text_dims[2]-offset))
+        elseif text_align == :bottom
+            shift = ((0, text_dims[2]+offset), (0, text_dims[2]+offset))
+        elseif text_align == :left
+            shift = ((-text_dims[1]-offset, 0), (-text_dims[1]-offset, 0))
+        elseif text_align == :right
+            shift = ((text_dims[1]+offset, 0), (text_dims[1]+offset, 0))
+        end
```
Comment (Wikunia) [inline src/node_shapes.jl:123]: You have a lot of code copy paste i.e here line 78-88 are the same as here

File: src/node_shapes.jl [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+"""
+    node_text(text::AbstractString, offset::Point, color::String)
+    node_text(text::LaTeXString, offset::Point, color::String)
+
+Label a node with text, which is also rotated to match the direction specified by `offset`.
+
+# Arguments
+- `offset::Point`: Specify the offset of the text label with respect to the node center.
+    - The text is translated and rotated in the direction of the offset.
+- `color`: Used only when `align` is not `:inside`. Specifies an offset in the direction of `align`.
```
Comment (Wikunia) [inline src/node_shapes.jl:138]: `color` is independent of `align`

File: src/node_shapes.jl [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+end
+
+"""
+    node_fill(fill::Symbol, arg::String)
+
+Fill in the interior of the node with an image or a color.
+
+# Arguments
+- `fill::Symbol`: Can be `image` or `color`.
+- `arg::String`: Can be a color like "red" or a image path in the same directory like "image.png".
```
Comment (Wikunia) [inline src/node_shapes.jl:171]: In Luxor one normally sets colors with symbols so `:red` instead of "red"

File: src/node_shapes.jl [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+Label a node with text or latex.
+
+# Arguments
+- `align::Symbol`: Specify the alignment of the text label with respect to the node.
+    - This position is determined with respect to the node `bounding_box` returned as a drawing option by [`node_shape`](@ref).
+    - Can take the values `:inside`, `:top`, `:bottom`, `:left`, `:right`. Default is `:inside`.
+- `color`: The color of the text. Default is "black".
+- `offset`: Used only when `align` is not `:inside`. Specifies an offset in the direction of `align`.
+"""
+function node_text(text::AbstractString, align::Symbol, color::String="black", offset::Real=1)
```
Comment (Wikunia) [inline src/node_shapes.jl:57]: I think we want to have an option to specify the font size?

File: src/node_shapes.jl [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+
+Create a border around the node shape.
+
+If the `:shape` is not specified, the `bounding_box` becomes the border.
+
+# Arguments
+- `color::Symbol`
+- `border_width::Real`: The line width used for the border. Defualt is 3.
+"""
+function node_border(color::String, border_width::Real=3)
```
Comment (Wikunia) [inline src/node_shapes.jl:208]: There is a mismatch in the `color` type here. You sometimes write `Symbol` and sometimes `String`. I think we shouldn't have any type here as we want to allow all kinds of colors.

**Review by Wikunia — CHANGES_REQUESTED** [review]
Additionally to the specific comments in the code:
You often restrict the types of arguments in your functions where it isn't necessary and sometimes is restricting the user.
Furthermore you can restructure parts of your functions to call helper functions to limit code copy pasting.

File: src/structs/GraphEdge.jl [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+)
+    g = graph.meta
+    if !(typeof(g) <: JGraph)
+        throw(ErrorException("Cannot define edge since $(typeof(graph)) is not a `JGraph`"))
+    end
+    if get_prop(g.adjacency_list, from_node, to_node) !== nothing
+        @warn "Edge $(from_node)=>$(to_node) is already created on canvas. Recreating it will leave orphan edges in the animation. To undo, call `rem_edge!`"
+    end
+    add_edge!(g.adjacency_list.graph, from_node, to_node)
+    set_prop!(g.adjacency_list, from_node, to_node, length(g.ordering)+1)
```
[the remark this answers is not in the document]
Comment (codejaeger) [reply src/structs/GraphEdge.jl:78]: `set_prop!` sets the property of the node/edge (edge in this case). Since the `WeightedGraph` type can only have one  property for each node and edge there is no name given to that property.

File: src/structs/GraphVertex.jl [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        if get_prop(g.adjacency_list, node) !== nothing
+            @warn "Node $(node) is already created on canvas. Recreating it will leave orphan node objects in the animation. To undo, call `rem_node!`"
+        end
+        if g.layout != :none
+            draw_fn = (args...; position=O, kwargs...) -> begin Luxor.translate(position); draw(args...; position=position, kwargs...) end
+        else
+            draw_fn = draw
+        end
+        add_vertex!(g.adjacency_list.graph)
+        set_prop!(g.adjacency_list, nv(g.adjacency_list), length(g.ordering)+1)
```
[the remark this answers is not in the document]
Comment (codejaeger) [reply src/structs/GraphVertex.jl:89]: Yes, to store more than one property for nodes and edges there is already the `MetaGraph` package. In many graph algorithms, only a single weight is associated with the nodes and edges hence this would be a useful structure to have.

File: src/structs/JGraph.jl (hunk 5 above)
Comment (codejaeger) [reply src/structs/JGraph.jl:199]: Yes.

File: src/node_shapes.jl (hunk 11 above)
Comment (codejaeger) [reply src/node_shapes.jl:171]: Hey, I was using `sethue` to switch between colors. I checked the methods available for `sethue` and could not find one using `Symbol`s. Is there something else you are referring to here?

Comment (Wikunia) [reply src/node_shapes.jl:171]: Oh sorry I think I got confused 😁 but I think other color packages are also supported. Let me check though

File: src/node_shapes.jl (hunk 12 above)
Comment (codejaeger) [reply src/node_shapes.jl:57]: To specify font and size, it is possible to add a (args...; kwargs...) -> setfont(<font>, <size>) anywhere in the list before a `node_text` is created. Since these functions are supposed to be backup options if the user wants to skip some drawing properties I skipped that option. Do you think it is still needed?

Comment (Wikunia) [reply src/node_shapes.jl:57]: I think font size is something so standard that we want to include it in `node_text`

Comment (codejaeger) [reply src/node_shapes.jl]: Okay sure will do that.

File: src/node_shapes.jl (hunk 11 above)
Comment (Wikunia) [reply src/node_shapes.jl]: See: https://juliagraphics.github.io/Luxor.jl/stable/colors-styles/

**Review by Wikunia — COMMENTED** [review]
Just some minor things

File: src/Javis.jl [hunk 16]
```diff
@@ -421,6 +437,27 @@ end
...
+    has_vertex,
+    has_edge,
+    add_vertex!,
+    rem_vertex!,
+    add_edge!,
+    rem_edge!,
+    weights,
+    node_props,
+    edge_props
+export node_shape, node_text, node_fill, node_border
```
Comment (Wikunia) [inline src/Javis.jl:460]: `edge_shape` is missing

File: src/structs/JGraph.jl [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+    height::Int;
+    mode::Symbol = :static,
+    layout::Symbol = :spring,
+    get_node_attribute::Function = (args...) -> nothing,
+    get_edge_attribute::Function = (args...) -> nothing,
+)
+    # Check available layouts
+    if !(layout in [:none, :spring, :spectral])
+        layout = :spring
+        @warn "Unknown layout '$(layout)'. Defaulting to static layout"
```
Comment (Wikunia) [inline src/structs/JGraph.jl:110]: do you mean: Defaulting to `:spring` layout?

File: examples/graph_animations/demo.md [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+
+function ground(args...) 
+    background("white")
+    sethue("black")
+end
+
+video = Video(400, 400)
+Background(1:150, ground)
+
+g = @Object 1:150 JGraph(true, 100, 100) O
```
Comment (Wikunia) [inline examples/graph_animations/demo.md:28]: Seems like the `O` doesn't have any effect?
For example replacing it by `Point(-100,0)` does not shift it 100px to the left

File: src/structs/JGraph.jl [hunk 19]
```diff
@@ -53,6 +53,10 @@ JGraph(directed::Bool, width::Int, height::Int) =
     directed ? JGraph(WeightedGraph(LightGraphs.SimpleDiGraph()), width, height) :
     JGraph(WeightedGraph(LightGraphs.SimpleGraph()), width, height)
 
+JGraph(directed::Bool, width::Int, height::Int, layout::Symbol=:none) =
```
Comment (Wikunia) [inline src/structs/JGraph.jl:56]: mode is still missing, right?

File: examples/graph_animations/demo2.md [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+
+function edge(; p1, p2, kwargs...)
+    sethue("black")
+    line(p1, p2, :stroke)
+end
+
+function node(position)
+    # Just return the node position
+    return Dict(:position=>position), (args...; kwargs...) -> nothing
+end
```
Comment (Wikunia) [inline examples/graph_animations/demo2.md:12]: this notation seems a bit weird. What are the two return arguments?
